src/controller/java/CHIPDeviceController-JNI.cpp

```diff
@@ -195,11 +195,36 @@ JNI_METHOD(jint, setNOCChain)
     jbyteArray operationalCertificate = (jbyteArray) env->CallObjectMethod(controllerParams, getOperationalCertificate);
     VerifyOrReturnValue(operationalCertificate != nullptr, CHIP_ERROR_BAD_REQUEST.AsInteger());
 
+    // use ipk and adminSubject from CommissioningParameters if not set in ControllerParams
+    CommissioningParameters commissioningParams = wrapper->GetCommissioningParameters();
+
+    Optional<Crypto::AesCcm128KeySpan> ipkOptional;
+    uint8_t ipkValue[CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES];
+    Crypto::AesCcm128KeySpan ipkTempSpan(ipkValue);
+
     jbyteArray ipk = (jbyteArray) env->CallObjectMethod(controllerParams, getIpk);
-    VerifyOrReturnValue(ipk != nullptr, CHIP_ERROR_BAD_REQUEST.AsInteger());
+    if (ipk != nullptr)
+    {
+        JniByteArray jByteArrayIpk(env, ipk);
+
+        VerifyOrReturnValue(jByteArrayIpk.byteSpan().size() == sizeof(ipkValue), CHIP_ERROR_INTERNAL.AsInteger());
+        memcpy(&ipkValue[0], jByteArrayIpk.byteSpan().data(), jByteArrayIpk.byteSpan().size());
+
+        ipkOptional.SetValue(ipkTempSpan);
+    }
+    else if (commissioningParams.GetIpk().HasValue())
+    {
+        // if no value pass in ControllerParams, use value from CommissioningParameters
+        ipkOptional.SetValue(commissioningParams.GetIpk().Value());
+    }
 
     Optional<NodeId> adminSubjectOptional;
     uint64_t adminSubject = env->CallLongMethod(controllerParams, getAdminSubject);
+    if (adminSubject == kUndefinedNodeId)
+    {
+        // if no value pass in ControllerParams, use value from CommissioningParameters
+        adminSubject = commissioningParams.GetAdminSubject().ValueOr(kUndefinedNodeId);
```

Comment:
Not having a value should be an error, since otherwise the admin subject will be set to an invalid value down the line

Reply:
Ok, but the OnDeviceNOCChainGeneration takes an Optional adminSubject and handles that case by substituting the nodeId of the controller so it didn't seem right to force the value to be set here since other places in the code allow it not to be set. 

void DeviceCommissioner::OnDeviceNOCChainGeneration(void * context, CHIP_ERROR status, const ByteSpan & noc, const ByteSpan & icac,
                                                    const ByteSpan & rcac, Optional<AesCcm128KeySpan> ipk,
                                                    Optional<NodeId> adminSubject)

Reply:
added comment